clippy_lints/src/methods/option_map_unwrap_or.rs

```diff
@@ -56,24 +64,30 @@ pub(super) fn lint<'a, 'tcx>(
              This can be done more directly by calling `{}` instead",
             arg, suggest
         );
-        // lint, with note if neither arg is > 1 line and both map() and
-        // unwrap_or() have the same span
-        let multiline = map_snippet.lines().count() > 1 || unwrap_snippet.lines().count() > 1;
-        let same_span = map_args[1].span.ctxt() == unwrap_args[1].span.ctxt();
-        if same_span && !multiline {
-            let suggest = if unwrap_snippet == "None" {
-                format!("and_then({})", map_snippet)
-            } else {
-                format!("map_or({}, {})", unwrap_snippet, map_snippet)
-            };
-            let note = format!(
-                "replace `map({}).unwrap_or({})` with `{}`",
-                map_snippet, unwrap_snippet, suggest
+
+        span_lint_and_then(cx, OPTION_MAP_UNWRAP_OR, expr.span, msg, |db| {
+            let map_arg_span = map_args[1].span;
+
+            db.multipart_suggestion(
+                &format!("use `{}` instead", suggest),
+                vec![
+                    (
+                        map_span,
+                        String::from(if unwrap_snippet_none { "and_then" } else { "map_or" }),
+                    ),
+                    (expr.span.with_lo(unwrap_args[0].span.hi()), String::from("")),
+                    (
+                        map_arg_span.with_hi(map_arg_span.lo()),
+                        if unwrap_snippet_none {
+                            String::from("")
```

Comment:
Doesn't this replace an empty span with an empty String? I'm not sure I'm a fan of that. Wouldn't it be better to create the `Vec` first and then conditionally `push` the element?